mirrord/agent/src/cgroup.rs

```diff
@@ -0,0 +1,262 @@
+//! Logic for pausing using cgroup directly
+//! Pause requires privileged - assumes ephemeral (hardcoded pid 1)
+use std::path::{Path, PathBuf};
+
+use nix::mount::{mount, MsFlags};
+use tokio::{
+    fs::{create_dir, try_exists, File, OpenOptions},
+    io::{AsyncBufReadExt, AsyncWriteExt, BufReader},
+};
+use tracing::{trace, warn};
+
+use crate::{
+    error::{AgentError, Result},
+    namespace::{set_namespace, NamespaceType},
+};
+
+const CGROUP_MOUNT_PATH: &str = "/mirrord_cgroup";
+const CGROUP_SUBGROUP_MOUNT_PATH: &str = "/mirrord_cgroup/subgroup";
+const CGROUPV2_PROCS_FILE: &str = "cgroup.procs";
+
+#[derive(Debug)]
+pub(crate) struct CgroupV1 {
+    /// Path to the the process' cgroup - usually /kubepods/besteffort/ID
+    cgroup_path: PathBuf,
+}
+
+impl CgroupV1 {
+    #[tracing::instrument(level = "trace", ret)]
+    pub(crate) async fn new() -> Result<Self> {
+        let file = File::open("/proc/1/cgroup")
+            .await
+            .map_err(|_| AgentError::PauseFailedCgroup("opening proc cgroup failed".to_string()))?;
```

Comment:
We're using this `AgentError::PauseFailedCgroup` with `map_err` a lot, due to this error being a kind of "catch all" and having the context be a string.

One of the reasons for using `thiserror` instead of `anyhow` is to have stronger error types and avoid this "string as context".

Maybe this error should in the category of `PauseCgroup::OpeningProc`.